### DISCUSSION TITLE
Cannot import extensionless

### DISCUSSION THREAD

**Discussion by shir0u** [TypeStrong/ts-node#1737] [General]
If I include the `.ts` extension, typescript will complain:

```
TS2691: An import path cannot end with a '.ts' extension
```

If I don't include the `.ts` extension, ts-node will complain:

```
throw new ERR_MODULE_NOT_FOUND(
          ^
CustomError: Cannot find module
```

Why is this? Why does it work fine for tsc but not ts-node?

**Comment by cspotcode** [comment]
It doesn't work fine for tsc though, since you get that error.

https://devblogs.microsoft.com/typescript/announcing-typescript-4-7-beta/#type-in-package-json-and-new-extensions

> Relative import paths need full extensions (we have to write `import "./foo.js"` instead of `import "./foo"`).

TSC's rule is that it has to be the *emitted* extension, so it has to be `.js`.  ts-node does the right thing to make this work.

**Reply by shir0u** [reply]
But tsc (4.6.3) compiles fine though and understands that the extensionless means .ts in ts file, and when compiled will append .js extension.

I see that you mentioned a beta version, that's good to know for future proof. Thanks I got the answer now 👍

**Reply by cspotcode** [reply]
Neither of those are correct, actually:

a) tsc (4.6.3) does not append the .js extension.

b) the behavior I described predates the beta

You can prove this for yourself.  Put a variety of imports in a .ts file, compile them, and check the output .js file.  See if any extensions have been added, removed, or rewritten.

**Reply by shir0u** [reply]
hmm maybe I was being confused with a bundler instead of using tsc directly. Thanks for the prompt reply and correction, appreciate it.
